INSTALL.md

```diff
@@ -146,9 +154,6 @@ be edited here.
         ### exit psql
         smokealarm_development=# \q        
 
-        ### Run sequelize migrations
```

Comment:
So a "migration" is, in general, a thing that updates an earlier schema to a later schema, but under the assumption that there is no data in place that also needs conversion or upgrading?

(I guess this is a question about how the word "migration" is generally used in this context, and also about how we specifically are using that word.)

Reply:
I think that generally, and ideally for this project going forward, that migration also includes data migration that needs to happen in an automated fashion.  However, it shouldn't include development specific data migration or population.  The migrations that are run as part of the INSTALL.md don't seem to do a great job of walking that line correctly.

Reply:
We can definitely go over this in a call, but for posterity's sake, I'll put it here too.

So, there's "ideal project situation" and "smoke alarm portal" situation.  I'll do ideal first.

# Ideal situation

> * How does the schema get set up when developing from scratch?

You do migrations from day one.  When you add a table, migration.  When you drop a column, migration.  That way you can always do upgrades/downgrades as development needs (switch branches, features, etc).  Also these days migrations are a really easy to write, with the framework taking your "addColumn" command, generating the sql for both upgrade, and downgrade, and doing extra checking for you.

> * What kinds of migrations do we have, and when do we run them? Is running migrations part of the normal dev setup? Do they just set up schema, or do they also migrate data?

You have migrations for development data, that are marked as development only migrations, and are run at the time that inserting that data is accurate.  Then further migrations that modify the tables are run after the insertion migrations.  If you need to do updates to do columns specific to the development data, you write development specific updates on keys that are in the original development only insertions, marking these migrations also as dev only.

However, you can also optionally have a development dump of a database that includes the metadata (that all migration systems have) of what migration was last run to generate that database.  So then you're method is "import sql file, run migrations" and any migrations that were added after that sql file was generated will be run.  You can update those sql files at your leisure.

> * Do migrations ever get run on prod, or do we assume that prod is always manually migrated?

Migrations always get run on prod.  This is the way that the database gets updated for production as it is in source control, documented, and tested.  Indeed, it's regression tested every time someone builds the database from nothing using only the migrations to boot it.  Development only migrations don't get run.

Not only that, but you get, for free, what version of the schema production is currently on.  You have to track that somewhere, or else how will you know what sql to manually run when you decide to put in features A..X into production.

# smoke-alarm-portal situation

> * How does the schema get set up when developing from scratch?

Before this pull, there was no sql file, and no migrations.  It was just sequelize evaluating the object model in models/index.js and assocaited files, and then generating a schema to match.  That schema needed to match the data files that followed.

After this pull, there will be a set sql file that holds the base schema, and then migrations that will run after it.  I did it this way because creating migrations from the ground up for the schema is too much of a scope creep for this project.  This works, and because the team is so small and development so rare, will suffice for now.

> * What kinds of migrations do we have, and when do we run them? Is running migrations part of the normal dev setup? Do they just set up schema, or do they also migrate data?

After this pull, migrations would be run as normal development setup after the base sql is run.  We would then run migrations all the time.  When switching to a new feature branch that has them, you should run them before using/working on it to get your database in line with the model.  When switching, you need to back out the migrations before using the other feature branch.

As for the data, maybe we should change it so that the sql file incorporates development data at a given timestamp of migration?  I could be convinced.  I was trying to do as few changes as possible, but it would definitely be some tech debt removal to have a "base schema + development data" sql file that could be loaded before migrations were run.  That could be updated as needed.

Right now there are about 6 different development data sql files that can optionally be run depending on the developer needs.  There would have to be a decision about which ones were included, or removed entirely.

> * Do migrations ever get run on prod, or do we assume that prod is always manually migrated?

I would like to start, because it's a lot more straightforward than the old way of manually running sql, and is more maintainable.  But before now, it wasn't done.